libvirt/README.md

```diff
@@ -44,3 +44,222 @@ If you want to set default password for podvm debugging then you can use guestfi
 
 Refer to the [cloud-api-adaptor deployment instructions](../install/README.md#deploy-cloud-api-adaptor)
 * If your libvirt host is not configured to be passwordless make sure you set ssh-key-secret in the [kustomization file](../install/overlays/libvirt/kustomization.yaml)
+
+# Creating an end-to-end environment for testing and development
+
+In this section you will learn how to setup an environment in your local machine to run peer pods with
+the libvirt cloud API adaptor. Bear in mind that many different tools can be used to setup the environment
+and here we just make suggestions of tools that seems used by most of the peer pods developers.
+
+## Requirements
+
+You must have a Linux/KVM system with libvirt installed and the following tools:
+
+- docker (or podman-docker)
+- [kubectl](https://kubernetes.io/docs/reference/kubectl/)
+- [kcli](https://kcli.readthedocs.io/en/latest/)
+
+Assume that you have a 'default' network and storage pools created in libvirtd system instance (`qemu:///system`). However,
+if you have a different pool name then the scripts should be able to handle it properly.
+
+## Create the Kubernetes cluster
+
+Use the [`kcli_cluster.sh`](./kcli_cluster.sh) script to create a simple two VMs (one master and one worker) cluster
+with the kcli tool, as:
+
+```
+./kcli_cluster.sh create
+```
+
+With `kcli_cluster.sh` you can configure the libvirt network and storage pools that the cluster VMs will be created, among
+other parameters. Run `./kcli_cluster.sh -h` to see the help for further information.
+
+If everything goes well you will be able to see the cluster running after setting your Kubernetes config with:
+
+`export KUBECONFIG=$HOME/.kcli/clusters/peer-pods/auth/kubeconfig`
+
+For example, shown below:
+
+```
+$ kcli list kube
++-----------+---------+-----------+---------------------------------------+
+|  Cluster  |   Type  |    Plan   |                  Vms                  |
++-----------+---------+-----------+---------------------------------------+
+| peer-pods | generic | peer-pods | peer-pods-master-0,peer-pods-worker-0 |
++-----------+---------+-----------+---------------------------------------+
+$ kubectl get nodes
+NAME                 STATUS   ROLES                  AGE     VERSION
+peer-pods-master-0   Ready    control-plane,master   6m8s    v1.25.3
+peer-pods-worker-0   Ready    worker                 2m47s   v1.25.3
+```
+
+## Prepare the Pod VM volume
+
+In order to build the Pod VM without installing the build tools, you can use the Dockerfiles hosted on [../podvm](../podvm) directory to run the entire process inside a container.
+
+To build a Ubuntu VM image, run:
+
+```
+$ cd ../podvm
+$ docker build --no-cache -t podvm_builder -f Dockerfile.podvm_builder .
+$ docker build --no-cache -t podvm_libvirt --build-arg BUILDER_IMG=podvm_builder:latest \
```

Comment:
When running this I got an error in the libsecomp stage:
```
checking whether the C compiler works... no
configure: error: in `/tmp/build-libseccomp.Pv69G/libseccomp-2.5.4':
configure: error: C compiler cannot create executables
See `config.log' for more details
make: *** [Makefile.inc:100: /src/cloud-api-adaptor/podvm/staticlib/kata-libseccomp/lib/libseccomp.a] Error 77
```
I know it's not changed in the PR, but I thought it was worth noting.

Reply:
The ci-e2e_libvirt branch on my git tree misses a fix that I sent yesterday in another PR: https://github.com/confidential-containers/cloud-api-adaptor/pull/537

Please rebase your tree and then those commands should work out.

Reply:
Updated my tree.

Reply:
I hit a timeout in the image creation again which was quite upsetting after waiting ~1 hour for it to run:
```
==> qemu.ubuntu: Waiting for SSH to become available...
==> qemu.ubuntu: Timeout waiting for SSH.
==> qemu.ubuntu: Deleting output directory...
Build 'qemu.ubuntu' errored after 5 minutes 57 seconds: Timeout waiting for SSH.

==> Wait completed after 5 minutes 57 seconds

==> Some builds didn't complete successfully and had errors:
--> qemu.ubuntu: Timeout waiting for SSH.

==> Builds finished but no artifacts were created.
make: *** [Makefile:69: podvm-f080e93-dirty-amd64.qcow2] Error 1
```

I tried updating `qcow2/ubuntu/qemu-ubuntu.pkr.hcl` to have `ssh_wait_timeout  = "600s"`, but I'm not sure if that's the correct setting as it didn't seem to help and ` 5 minutes 57 seconds` is longer than the default of 300s?
```
==> qemu.ubuntu: Timeout waiting for SSH.
==> qemu.ubuntu: Deleting output directory...
Build 'qemu.ubuntu' errored after 6 minutes 31 seconds: Timeout waiting for SSH.

==> Wait completed after 6 minutes 31 seconds

==> Some builds didn't complete successfully and had errors:
--> qemu.ubuntu: Timeout waiting for SSH.

==> Builds finished but no artifacts were created.
make: *** [Makefile:69: podvm-f080e93-dirty-amd64.qcow2] Error 1
```

Reply:
It's worth noting that this passed for me in less than 4 mins when I ran it not in the container:
```
Build 'qemu.ubuntu' finished after 3 minutes 58 seconds.

==> Wait completed after 3 minutes 58 seconds
```

Reply:
@wainersm, @stevenhorsman , i didn't hit this issue but one thing that may improve it is to pass `--device=/dev/kvm` (or even --privileged ) as it seems the VM starting inside the container is not accelerated with kvm, in any case worth mentioning it i think
Thanks to @whitebrandy who noticed this!

Reply:
@snir911  hey Snir, sorry for being slow, but where do you mean to add `--device=/dev/kvm`. It looks like `docker build`, doesn't take that as an argument, just `docker run`?
```
podvm# docker build --no-cache --device=/dev/kvm -t podvm_builder -f Dockerfile.podvm_builder .
unknown flag: --device
See 'docker build --help'.
```

Reply:
oh, podman does, so i took it for granted also docker :P

Reply:
So good news and bad news on this I guess: I still can't get docker to work and the suggested flags don't exist, but I switched to podman (and added `--device=/dev/kvm`) and it all ran without an issue. 

I guess the question is what to do here:
- Change to documentation to require podman
- Add a note to say that if docker fails to try pdoman
- Try something else to get docker working

Reply:
As another data point - I took a fresh VM and running podman without `--device=/dev/kvm` also fails to build the image, so on my set-up (4 vCPU, 8GB VM) it seems like that is mandatory to get success here.

Reply:
If you don't pass /dev/kvm then QEMU will run on emulated (TCG) mode, which is slower than hardware virtualization (KVM).  We cannot use KVM on cross-build nor on github jobs (e.g. https://github.com/confidential-containers/cloud-api-adaptor/blob/staging/.github/workflows/podvm.yaml) because Github default runners don't support nested virt.

We can try to speed up the emulated mode. For example, use `q35` instead of `pc` machine type (https://github.com/confidential-containers/cloud-api-adaptor/blob/staging/podvm/qcow2/ubuntu/variables.pkr.hcl#L56). And/or increase a little bit the requested memory (2GB) - https://github.com/confidential-containers/cloud-api-adaptor/blob/staging/podvm/qcow2/ubuntu/variables.pkr.hcl#L31

On the podvm/README.md (see https://github.com/confidential-containers/cloud-api-adaptor/issues/550) document the environment constraints and required minimum memory/disk.

I don't like requiring podman. It seems more of our developers run docker.

Anyway, doesn't things shouldn't be addressed on this PR. @stevenhorsman @snir911 do you think it is good enough to get merged?

Reply:
So for this particular issue of getting the podvm built in a container, I'm ok for the tidy up to be added to part of #550.

For the wider PR, I'm a bit more mixed. - I think it's a great start, but the fact that I've not been able to get the scripts working and produce a libvirt peer pod set-up does bother me. It suggests that there is either a problem in the code (unlikely), a problem in what I'm doing (likely, but that indicates there is possibly a documentation problem), or a issue in my set-up that is different (again not something that is probably not documented). My concern with merging it so that we can then build on it is if we'll end up in a situation that no-one outside of RedHat can get it working, which isn't a great place. If it is just me that can't get it working, then I'm fine with it though.

On the other hand if we can merge then, then start automating and improving some of the steps e.g. with ansible scripts, and get a libvirt CI/CD running in the next step after this in open-source then that would be a huge step and worth 'breaking some eggs' for.

Reply:
@stevenhorsman I'm pretty sure it would work for you with more resources, for sure if it runs on baremetal and not in a VM (as the VM packer uses is nested and in addition it's not accelerated) . Maybe we can have a note mentioning this intensive operation that requires a lot of resources and shouldn't run in a VM or something.
BTW when i tried to run the build for RHEL in a VM it failed due to disk size, I'm currently trying to optimize it on this aspect (https://github.com/confidential-containers/cloud-api-adaptor/pull/525) IDK if we can improve it on the other aspects as well

Reply:
It seems emulating x86_64 on m1 has lower performanc,. I think it worth trying 8vCPUs / 32GB if it's not a burden

Reply:
FYI - the podvm build with docker worked fine in my 8 vCPU, 32GB RAM IBM Cloud VSI

Reply:
I tried within a VM with 4 vCPUs, 8GB Mem, and 30GB disk on my Linux Fedora 34 laptop. It worked out.

Of course the build still requires much disk (being addressed in https://github.com/confidential-containers/cloud-api-adaptor/pull/525) and packer within container takes a good amount of time because QEMU emulated. This last problem can be mitigated by passing /dev/kvm (i.e. allowing KVM acceleration on container). I will be documenting all this on the PR for https://github.com/confidential-containers/cloud-api-adaptor/issues/550

Reply:
> @wainersm, @stevenhorsman , i didn't hit this issue but one thing that may improve it is to pass `--device=/dev/kvm` (or even --privileged ) as it seems the VM starting inside the container is not accelerated with kvm, in any case worth mentioning it i think Thanks to @whitebrandy who noticed this!

So, back to the original comment, the closest I know of is specifying "RUN --security=insecure" which is similar to running "docker run --privileged" but sorry, untested :)